File: services/graphql-server/src/services/switch.js
const grpc = require('grpc')
const opentracing = require('opentracing')

const { proto } = require('../grpc')
const Logger = require('../utils/logger')
const { createTracer } = require('../utils/tracer')

class SwitchService {
  constructor (config, options) {
    options = options || {}
    this.serviceAddr = config.switchServiceAddr
    this.logger = options.logger || new Logger('SwitchService')
    this.histogram = options.histogram || { observe () {} }
    this.summary = options.summary || { observe () {} }
    this.tracer = options.tracer || createTracer({ serviceName: 'SwitchService' })
    this.client = options.client || new proto.SwitchService(
      this.serviceAddr,
      grpc.credentials.createInsecure()
    )
  }

  async exec (context, name, func) {
    let result, err
    let latency

    // https://opentracing-javascript.surge.sh/interfaces/spanoptions.html
    // https://github.com/opentracing/specification/blob/master/semantic_conventions.md
    const span = this.tracer.startSpan(name, { childOf: context.span }) // { childOf: context.span.context() }
    span.setTag(opentracing.Tags.SPAN_KIND, 'client')
    span.setTag(opentracing.Tags.PEER_SERVICE, 'switch-service')
    span.setTag(opentracing.Tags.PEER_ADDRESS, this.serviceAddr)

    const carrier = {}
    this.tracer.inject(span, opentracing.FORMAT_TEXT_MAP, carrier)

    const metadata = new grpc.Metadata()
    metadata.add('span.context', JSON.stringify(carrier))

    // Core functionality
    try {
      const startTime = Date.now()
      result = await func(metadata)
      latency = (Date.now() - startTime) / 1000
    } catch (e) {
      err = e
      this.logger.error(err)
    }

    // Metrics
    const labelValues = { op: name, success: err ? 'false' : 'true' }
    this.histogram.observe(labelValues, latency)
    this.summary.observe(labelValues, latency)

    // Tracing
    span.log({
      event: name,
      message: err ? err.message : 'successful!'
    })
    span.finish()

    if (err) {
      throw err
    }

    return result
  }

  installSwitch (context, { siteId, name, state, states }) {
    return this.exec(context, 'install-switch', metadata => {
      return new Promise((resolve, reject) => {
        this.client.installSwitch({ siteId, name, state, states }, metadata, (err, swtch) => {
          return err ? reject(err) : resolve(swtch)
        })
      })
    })
  }

  removeSwitch (context, id) {
    return this.exec(context, 'remove-switch', metadata => {
      return new Promise((resolve, reject) => {
        this.client.removeSwitch({ id }, metadata, err => {
          return err ? reject(err) : resolve()
        })
      })
    })
  }

  getSwitch (context, id) {
    return this.exec(context, 'get-switch', metadata => {
      return new Promise((resolve, reject) => {
        this.client.getSwitch({ id }, metadata, (err, swtch) => {
          return err ? reject(err) : resolve(swtch)
        })
      })
    })
  }

  getSwitches (context, siteId) {
    return this.exec(context, 'get-switches', metadata => {
      return new Promise((resolve, reject) => {
        const switches = []
        const call = this.client.getSwitches({ siteId }, metadata)
        call.on('data', swtch => switches.push(swtch))
        call.on('error', err => reject(err))
        call.on('end', () => resolve(switches))
      })
    })
  }

  setSwitch (context, id, { state }) {
    return this.exec(context, 'set-switch', metadata => {
      return new Promise((resolve, reject) => {
        this.client.setSwitch({ id, state }, metadata, err => {
          if (err) {
            reject(err)
          }

          // switch-service does not respond with updated switch
          this.client.getSwitch({ id }, metadata, (err, swtch) => {
            return err ? reject(err) : resolve(swtch)
          })
        })
      })
    })
  }
}

module.exports = SwitchService

File: services/switch/internal/service/switch.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/moorara/microservices-demo/services/switch/internal/metrics"
	"github.com/moorara/microservices-demo/services/switch/internal/model"
	"github.com/moorara/microservices-demo/services/switch/internal/proto"
	"github.com/moorara/microservices-demo/services/switch/pkg/log"
	"github.com/opentracing/opentracing-go"
	"github.com/opentracing/opentracing-go/ext"
	"google.golang.org/grpc/metadata"

	arango "github.com/arangodb/go-driver"
	opentracingLog "github.com/opentracing/opentracing-go/log"
)

const queryGetSwitches = `FOR sw IN switches FILTER sw.siteId == @siteId RETURN sw`

type (
	callback func() error

	// SwitchService implements proto.SwitchServiceServer
	SwitchService struct {
		arango  ArangoService
		logger  *log.Logger
		metrics *metrics.Metrics
		tracer  opentracing.Tracer
	}
)

// NewSwitchService creates a new switch service
func NewSwitchService(arango ArangoService, logger *log.Logger, metrics *metrics.Metrics, tracer opentracing.Tracer) proto.SwitchServiceServer {
	return &SwitchService{
		arango:  arango,
		logger:  logger,
		metrics: metrics,
		tracer:  tracer,
	}
}

func (s *SwitchService) extractParentSpanContext(ctx context.Context) (opentracing.SpanContext, error) {
	meta, ok := metadata.FromIncomingContext(ctx)
	if ok {
		vals := meta.Get("span.context")
		if len(vals) == 1 {
			data := make(map[string]string)
			err := json.Unmarshal([]byte(vals[0]), &data)
			if err != nil {
				return nil, err
			}

			carrier := opentracing.TextMapCarrier(data)
			return s.tracer.Extract(opentracing.TextMap, carrier)
		}
	}

	return nil, nil
}

func (s *SwitchService) exec(ctx context.Context, req interface{}, op, query string, fn callback) {
	var span opentracing.Span
	parentSpanContext, _ := s.extractParentSpanContext(ctx)
	if parentSpanContext == nil {
		span = s.tracer.StartSpan(op)
	} else {
		span = s.tracer.StartSpan(op, opentracing.ChildOf(parentSpanContext))
	}

	// https://github.com/opentracing/specification/blob/master/semantic_conventions.md
	defer span.Finish()
	ext.DBType.Set(span, "arango")
	ext.DBStatement.Set(span, query)
	span.LogFields(opentracingLog.String("event", op))

	start := time.Now()
	err := fn()
	latency := time.Since(start).Seconds()

	success := "true"
	if err != nil {
		success = "false"
		s.logger.Error("message", fmt.Sprintf("%s failed: %s", op, err))
		span.LogFields(opentracingLog.String("message", err.Error()))
	} else {
		s.logger.Debug("message", fmt.Sprintf("%s succeeded.", op), "req", req)
		span.LogFields(opentracingLog.String("message", "successful!"))
	}

	s.metrics.OpLatencyHist.WithLabelValues(op, success).Observe(latency)
	s.metrics.OpLatencySumm.WithLabelValues(op, success).Observe(latency)
}

// InstallSwitch creates a new switch
func (s *SwitchService) InstallSwitch(ctx context.Context, req *proto.InstallSwitchRequest) (*proto.Switch, error) {
	var err error
	var meta arango.DocumentMeta

	doc := &model.Switch{
		Key:    uuid.New().String(),
		SiteID: req.GetSiteId(),
		Name:   req.GetName(),
		State:  req.GetState(),
		States: req.GetStates(),
	}

	s.exec(ctx, req, "InstallSwitch_CreateDocument", "CreateDocument", func() error {
		meta, err = s.arango.CreateDocument(ctx, doc)
		return err
	})

	if err != nil {
		return nil, err
	}

	doc.ID = meta.ID.String()
	doc.Key = meta.Key
	doc.Rev = meta.Rev

	return &proto.Switch{
		Id:     meta.Key,
		SiteId: doc.SiteID,
		Name:   doc.Name,
		State:  doc.State,
		States: doc.States,
	}, nil
}

// RemoveSwitch deletes a switch
func (s *SwitchService) RemoveSwitch(ctx context.Context, req *proto.RemoveSwitchRequest) (*proto.RemoveSwitchResponse, error) {
	var err error
	key := req.GetId()

	s.exec(ctx, req, "RemoveSwitch_RemoveDocument", "RemoveDocument", func() error {
		_, err = s.arango.RemoveDocument(ctx, key)
		return err
	})

	if err != nil {
		return nil, err
	}

	return &proto.RemoveSwitchResponse{}, nil
}

// GetSwitch retrieves a switch
func (s *SwitchService) GetSwitch(ctx context.Context, req *proto.GetSwitchRequest) (*proto.Switch, error) {
	var err error
	key := req.GetId()
	doc := &model.Switch{}

	s.exec(ctx, req, "GetSwitch_ReadDocument", "ReadDocument", func() error {
		_, err = s.arango.ReadDocument(ctx, key, doc)
		return err
	})

	if err != nil {
		return nil, err
	}

	return &proto.Switch{
		Id:     doc.Key,
		SiteId: doc.SiteID,
		Name:   doc.Name,
		State:  doc.State,
		States: doc.States,
	}, nil
}

// GetSwitches retrieves a group of switches
func (s *SwitchService) GetSwitches(req *proto.GetSwitchesRequest, stream proto.SwitchService_GetSwitchesServer) error {
	var err error
	var cursor arango.Cursor

	ctx := stream.Context()
	vars := map[string]interface{}{
		"siteId": req.GetSiteId(),
	}

	s.exec(ctx, req, "GetSwitches_Query", queryGetSwitches, func() error {
		cursor, err = s.arango.Query(ctx, queryGetSwitches, vars)
		return err
	})

	if err != nil {
		return err
	}

	defer cursor.Close()

	s.exec(ctx, req, "GetSwitches_ReadDocument_Send", "ReadDocument", func() error {
		for cursor.HasMore() {
			doc := &model.Switch{}
			_, err = cursor.ReadDocument(ctx, doc)
			if err != nil {
				return err
			}

			sw := &proto.Switch{
				Id:     doc.Key,
				SiteId: doc.SiteID,
				Name:   doc.Name,
				State:  doc.State,
				States: doc.States,
			}

			err = stream.Send(sw)
			if err != nil {
				return err
			}
		}

		return nil
	})

	return err
}

// SetSwitch changes the state of a switch
func (s *SwitchService) SetSwitch(ctx context.Context, req *proto.SetSwitchRequest) (*proto.SetSwitchResponse, error) {
	var err error
	key := req.GetId()
	doc := &model.Switch{
		State: req.GetState(),
	}

	s.exec(ctx, req, "SetSwitch_UpdateDocument", "UpdateDocument", func() error {
		_, err = s.arango.UpdateDocument(ctx, key, doc)
		return err
	})

	if err != nil {
		return nil, err
	}

	return &proto.SetSwitchResponse{}, nil
}
